File: cinema-manager/backend/src/main/java/com/example/cinema/managing/system/service/MerchandiseService.java
package com.example.cinema.managing.system.service;

import com.example.cinema.managing.system.model.Merchandise;
import com.example.cinema.managing.system.repository.MerchandiseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class MerchandiseService {

    @Autowired
    private MerchandiseRepository merchandiseRepository;

    public List<Merchandise> getAllMerchandise() {
        return merchandiseRepository.findAll();
    }

    public List<Merchandise> getActiveMerchandise() {
        return merchandiseRepository.findByActiveTrue();
    }

    public Optional<Merchandise> getMerchandiseById(String id) {
        return merchandiseRepository.findById(id);
    }

    public List<Merchandise> getMerchandiseByCategory(String category) {
        return merchandiseRepository.findByCategoryAndActiveTrue(category);
    }

    public List<Merchandise> getBundleMerchandise() {
        return merchandiseRepository.findByBundleTrue();
    }

    public Merchandise createMerchandise(Merchandise merchandise) {
        merchandise.setCreatedAt(LocalDateTime.now());
        merchandise.setSalesCount(0);
        return merchandiseRepository.save(merchandise);
    }

    public Merchandise updateMerchandise(String id, Merchandise merchandiseDetails) {
        Merchandise merchandise = merchandiseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Merchandise not found with id: " + id));
        
        merchandise.setName(merchandiseDetails.getName());
        merchandise.setDescription(merchandiseDetails.getDescription());
        merchandise.setPrice(merchandiseDetails.getPrice());
        merchandise.setCategory(merchandiseDetails.getCategory());
        merchandise.setImageUrl(merchandiseDetails.getImageUrl());
        merchandise.setStock(merchandiseDetails.getStock());
        merchandise.setBundle(merchandiseDetails.isBundle());
        merchandise.setBundleMovieId(merchandiseDetails.getBundleMovieId());
        merchandise.setActive(merchandiseDetails.isActive());
        
        return merchandiseRepository.save(merchandise);
    }

    public void deleteMerchandise(String id) {
        merchandiseRepository.deleteById(id);
    }

    public Merchandise updateStock(String id, Integer quantity) {
        Merchandise merchandise = merchandiseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Merchandise not found with id: " + id));
        
        merchandise.setStock(merchandise.getStock() + quantity);
        return merchandiseRepository.save(merchandise);
    }

    public void incrementSalesCount(String id) {
        Merchandise merchandise = merchandiseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Merchandise not found with id: " + id));
        
        merchandise.setSalesCount(merchandise.getSalesCount() + 1);
        merchandiseRepository.save(merchandise);
    }
}

File: cinema-manager/frontend/src/services/merchandiseService.js
import api from './api';

const merchandiseService = {
  // Get all active merchandise
  getAllMerchandise: async () => {
    const response = await api.get('/merchandise');
    return response.data;
  },

  // Get all merchandise (admin only)
  getAllMerchandiseAdmin: async () => {
    const response = await api.get('/merchandise/all');
    return response.data;
  },

  // Get merchandise by ID
  getMerchandiseById: async (id) => {
    const response = await api.get(`/merchandise/${id}`);
    return response.data;
  },

  // Get merchandise by category
  getMerchandiseByCategory: async (category) => {
    const response = await api.get(`/merchandise/category/${category}`);
    return response.data;
  },

  // Get bundle merchandise
  getBundleMerchandise: async () => {
    const response = await api.get('/merchandise/bundles');
    return response.data;
  },

  // Create new merchandise (admin only)
  createMerchandise: async (merchandiseData) => {
    const response = await api.post('/merchandise', merchandiseData);
    return response.data;
  },

  // Update merchandise (admin only)
  updateMerchandise: async (id, merchandiseData) => {
    const response = await api.put(`/merchandise/${id}`, merchandiseData);
    return response.data;
  },

  // Delete merchandise (admin only)
  deleteMerchandise: async (id) => {
    const response = await api.delete(`/merchandise/${id}`);
    return response.data;
  },

  // Update stock (admin only)
  updateStock: async (id, quantity) => {
    const response = await api.patch(`/merchandise/${id}/stock?quantity=${quantity}`);
    return response.data;
  }
};

export default merchandiseService;
